### PR TITLE
Xorless fee

### PR DESCRIPTION
**Pull request by ZlayaMorda**
* Added extrinsic to create xorless transactions
* extensions upgraded to implement xorless
* end session upgraded to implement xorless bbb
* tests added
* migration added
* config params updated

### CODE REVIEW

File: runtime/src/xor_fee_impls.rs [hunk 1]
```diff
@@ -425,17 +433,55 @@ impl xor_fee::WithdrawFee<Runtime> for WithdrawFee {
...
+    ) -> Result<
+        (
+            AccountId,
+            Option<NegativeImbalanceOf<Runtime>>,
+            Option<AssetId>,
+        ),
+        DispatchError,
+    > {
         match call {
+            // TODO: remake for xorless
```
Comment (ZlayaMorda) [inline runtime/src/xor_fee_impls.rs:445]: Would be another PR

File: pallets/xor-fee/src/lib.rs [hunk 2]
```diff
@@ -840,15 +1191,75 @@ pub mod pallet {
...
         }
+
+        /// Allow use assets from white list to pay for fee
+        /// # Parameters:
+        /// - `origin`: caller
+        /// - `call`: call for which pay fee
+        /// - `asset_id`: asset in which pay fee, where None - XOR
+        #[allow(unused_variables)] // Used in extension
+        #[pallet::call_index(3)]
+        #[pallet::weight(<T as Config>::WeightInfo::xorless_call())]
```
[the remark this answers is not in the document]
Comment (ZlayaMorda) [reply pallets/xor-fee/src/lib.rs:1202]: It is calculated inside extrinsic, and even if we can get weight of the provided `call` and use it in macro, this `call` may call another `call`, so it should be calculated like this

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -1306,13 +1306,19 @@ where
...
     pub const DEXIdValue: DEXId = 0;
+    pub const MaxWhiteListTokens: u32 = 30;
 }
 
 parameter_types! {
     pub const FeeReferrerWeight: u32 = 10; // 10%
     pub const FeeXorBurnedWeight: u32 = 20; // 20%
     pub const FeeValBurnedWeight: u32 = 50; // 50%
     pub const FeeKusdBurnedWeight: u32 = 20; // 20%
+    pub const MinimalFeeInAsset: Balance = ((FeeReferrerWeight::get()
```
[the remark this answers is not in the document]
Comment (ZlayaMorda) [reply runtime/src/lib.rs:1317]: It is minimal amount where may be calculated proportions. For swap it is used in minimal amount out, for remint_val value is zero. Also it is used for fee calculation to prevent values which cant be used in proportions. I don't if should and how should be calculated such value.

@vovac12 what do you think?

File: pallets/xor-fee/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (wer1st) [reply pallets/xor-fee/src/lib.rs:1202]: It will be better to estimate the weight before the calling, it is a reason why we have these weight annotation before extrinsics.
If some call contains another calls - it is better to sum all of them. But I'm not sure how can we do that. 
If we can get at least the weight of provided call - we have to do it.

File: runtime/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vovac12) [reply runtime/src/lib.rs:1317]: @wer1st as I understand this value is set as minimal amount which could be used in further calculations
@ZlayaMorda Maybe just set constant value instead of such calculation and explain it in comment?

[the remark this answers is not in the document]
Comment (ZlayaMorda) [reply runtime/src/lib.rs:1317]: Yeah, may set constant value, add comments where it's used

File: runtime/src/xor_fee_impls.rs [hunk 4]
```diff
@@ -425,17 +433,55 @@ impl xor_fee::WithdrawFee<Runtime> for WithdrawFee {
...
+                    None => {},
+                    Some(asset_id) if XorFee::whitelist_tokens().contains(&asset_id) => {
+                        let asset_fee = FixedWrapper::from(
+                            PriceTools::get_average_price(
+                                &GetXorAssetId::get(),
+                                &asset_id,
+                                PriceVariant::Buy)?
+                        ) * fee;
+                        let asset_fee = asset_fee.into_balance();
+                        if asset_fee.lt(&MinimalFeeInAsset::get()) {
```
Comment (ZlayaMorda) [inline runtime/src/xor_fee_impls.rs:466]: @vovac12 MinimalFeeInAsset used there as minimal amount for fee suitable for calculations.

File: runtime/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZlayaMorda) [reply runtime/src/lib.rs]: discussed and done

File: pallets/xor-fee/src/lib.rs [hunk 5]
```diff
@@ -193,73 +253,215 @@ where
...
+                            return Ok(());
+                        }
+                    }
+                    BurntForFee::<T>::mutate(asset_id, |balance| {
+                        balance.fee_without_referral = balance
+                            .fee_without_referral
+                            .saturating_add(corrected_fee_as_asset)
+                    });
+                }
+                #[cfg(not(feature = "wip"))] // Xorless fee
```
[the remark this answers is not in the document]
Comment (ZlayaMorda) [reply pallets/xor-fee/src/lib.rs:322]: done

**Review by vovac12 — APPROVED** [review]

**Review by shlavik — APPROVED** [review]
